design-proposals/klip-9-table-functions.md

```diff
@@ -0,0 +1,206 @@
+**Status**: In Discussion 
+**Discussion**: https://github.com/confluentinc/ksql/issues/527
+
+## Motivation and background
+
+## Background
+
+A _table valued function_ or _table generating function_ is a concept from the database world.
+It's a function that returns a set of zero or more rows. Contrast this to a scalar function which
+returns a single value. We will call them _table functions_ in the remainder of this document.
+
+Table functions are typically used in FROM clauses with traditional databases. Some databases
+such as PostgreSQL and Hive also allow table functions to appear in SELECT clauses.
+
+An example of a table function from Apache Hive is `explode` - this takes a non scalar type such as
+an ARRAY or MAP and explodes it to zero or more rows.
+
+`select explode(array('A','B','C'));`
+
+Would return three rows:
+
+```
+A
+B
+C
+```
+
+Table functions can be thought of as analagous to the `FlatMap` operation commonly found in 
+functional programming or stream processing frameworks such as Kafka Streams.
+
+Please see the following links for further background information:
+
+* [Apache Hive Table Generating Functions](https://cwiki.apache.org/confluence/display/Hive/LanguageManual+UDF#LanguageManualUDF-Built-inTable-GeneratingFunctions(UDTF)
+* [Microsoft SQL Server table functions](https://docs.microsoft.com/en-us/dotnet/framework/data/adonet/sql/linq/how-to-use-table-valued-user-defined-functions)
+* [Google BigQuery (see UNNEST)](https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax)
+* [PostgreSQL (see UNNEST)](http://www.postgresql.cn/docs/9.5/functions-array.html) 
+* [Kafka Streams (see FlatMap)](https://kafka.apache.org/20/documentation/streams/developer-guide/dsl-api.html)
+
+## Motivation
+
+We propose to support table functions in SELECT clauses in KSQL.
+
+There is [demand from KSQL users](https://github.com/confluentinc/ksql/issues/527) for a way to explode non scalar types such
+as ARRAY or MAP from columns in KSQL streams or tables into multiple rows. This is the main
+motivation for proposing this feature.
+
+Table funtions would also be useful for other cases such as splitting a string into words, or any
+case where you need to go from a single row to zero or more, for example, unaggregating data or
+duplicating rows.
+
+Here's an example of exploding an array from a column into multiple rows:
+
+We have a stream containing a batch of sensor readings from a particular sensor. For further
+stream processing we want to split them out into separate messages:
+
+Given:
+
+```
+{sensor_id:12345 readings: [23, 56, 3, 76, 75]}
+```
+
+We want to get:
+
+```
+{sensor_id:12345 reading: 23}
+{sensor_id:12345 reading: 56}
+{sensor_id:12345 reading: 3}
+{sensor_id:12345 reading: 76}
+{sensor_id:12345 reading: 75}
+```
+
+This could be accomplished by the following CSAS:
+
+```
+CREATE STREAM exploded_stream AS
+    SELECT sensor_id, EXPLODE(readings) AS reading FROM batched_readings;
```

Comment:
Supporting UDTFs in FROM clauses was my original suggestion and was discussed in the original thread here https://github.com/confluentinc/ksql/issues/527 - however, the consensus seemed to move on from there to UDTFs on the SELECT clause being a more pragmatic choice. 
I think it's fine to revisit this point, but I would like us to get some consensus on this as its such a fundamental point, and the prototype was based on the assumption that UDTFs on SELECTs is the way to go :)

Reply:
>> `SELECT x.*, f.*, i.* FROM x LATERAL VIEW finiteUdft() AS f LATERAL VIEW infinitieUdtf() As i;`

Presumably this would result in a cross join? I doubt that's what most users would want.

I made this point in an earlier comment, but I think zip semantics are going to be more in demand, as yet I haven't heard a single concrete use case where a cartesian join wold be useful. Here's the example of zip I used earlier:

```
{"country":"usa",customer_names:["tim", "derek", "apurva"], customer_ages[99, 98, 97]}
```

expanding to:

```
{"country":"usa", customer_name:"tim", customer_age:99}
{"country":"usa", customer_name:"derek", customer_age:98}
{"country":"usa", customer_name:"apurva", customer_age:97}
```

How would you express a zip using udtfs on the FROM clause?

I don't see how it could be done without either a) creating a new JOIN type (ZIP) b) Implementing correlated subqueries (unintuitive, hard to grok) c) doing a cartesian join and a separate filtering step in a separate query (inefficient O(N*N) plus poor latency due to intermediate topic)

Reply:
Having a SELECT that returns multiple rows per input row feels wrong. So, it's pretty interesting to see this is the route that Postgres has gone! 

If we're going this route than I agree that a zip join seems more intuitive in the select expression.  I guess we can always support cartesian product joins via other syntax later if we need.

It seems the main fly in the ointment now is the unordered nature of MAPs. I definitely feels wrong to have the result of two exploded MAPs be dependent on the iteration order of Java HashMap, which might mean it changes depending on Java version.

Reply:
> Perhaps we should disallow EXPLODE taking a MAP directly

Yeah, I came to the same conclusion when thinking about them later.  

> Then we could have another tableFunction SORTED_PAIRS(MAP) which takes a map and outputs an array of pairs sorted by key.

I think this can be a generic UDF that converts map -> array.

> If you have just one exploded MAP in a SELECT you probably don't care about order of the pairs so UNSORTED_PAIRS allows you to this without the O(NlogN) penalty of sorting.

On exploded map, _with no other UDTFs_ is fine, but as soon as you mix in another UDTF, regardless of param type, then the ordering becomes an issue.